### PR TITLE
PMM-4879 Adding support for defaults-file in new mysql service.

### PR DESCRIPTION
**Pull request by pkadej**
[PMM-4879](https://jira.percona.com/browse/PMM-4879)

Build: [SUBMODULES-2465](https://github.com/Percona-Lab/pmm-submodules/pull/2465)

- [pmm](https://github.com/percona/pmm/pull/928)

### CODE REVIEW

File: services/management/mysql.go [hunk 1]
```diff
@@ -78,6 +83,30 @@ func (s *MySQLService) Add(ctx context.Context, req *managementpb.AddMySQLReques
...
+				req.Password = result.Password
+			}
+
+			if len(result.Host) != 0 {
+				req.Address = result.Host
+			}
+
+			if result.Port > 0 {
+				req.Port = result.Port
+			}
```
Comment (BupycHuk) [inline services/management/mysql.go:107]: based on discussion in Jira ticket we should set values from defaults file only if they weren't passed as a separate field in request.

File: services/management/mysql.go [hunk 2]
```diff
@@ -39,15 +42,17 @@ type MySQLService struct {
 	state agentsStateUpdater
 	cc    connectionChecker
 	vc    versionCache
+	pfd   defaultsFileParser
```
Comment (BupycHuk) [inline services/management/mysql.go:45]:
```suggestion
	dfp   defaultsFileParser
```

**Review by BupycHuk — CHANGES_REQUESTED** [review]

File: services/agents/parse_defaults_file.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/percona/pmm/api/inventorypb"
+	"github.com/pkg/errors"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/utils/logger"
+)
+
+// ParseDefaultsFile requests from agent to parse defaultsFile.
+type ParseDefaultsFile struct {
```
[the remark this answers is not in the document]
Comment (pkadej) [reply services/agents/parse_defaults_file.go]: Sure, fixed.

File: services/agents/parse_defaults_file.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("ParseDefaultsFile response from agent: %+v.", resp)
+	parserResponse, ok := resp.(*agentpb.ParseDefaultsFileResponse)
+	if !ok {
+		return nil, errors.New("wrong response from agent (not ParseDefaultsFileResponse model)")
+	}
+	if len(parserResponse.GetError()) != 0 {
```
[the remark this answers is not in the document]
Comment (pkadej) [reply services/agents/parse_defaults_file.go]: Changed.

File: services/management/mysql.go [hunk 5]
```diff
@@ -78,6 +83,30 @@ func (s *MySQLService) Add(ctx context.Context, req *managementpb.AddMySQLReques
 		if err != nil {
 			return err
 		}
+
+		if len(req.DefaultsFile) != 0 {
```
[the remark this answers is not in the document]
Comment (pkadej) [reply services/management/mysql.go]: Done.

File: services/management/mysql.go (hunk 1 above)
Comment (pkadej) [reply services/management/mysql.go]: Changed.

File: services/management/mysql.go (hunk 2 above)
Comment (pkadej) [reply services/management/mysql.go]: Refactored.

**Comment by BupycHuk** [comment]
@pkadej could you please commit changes in go modules, so we can see result of CI.

**Comment by pkadej** [comment]
> @pkadej could you please commit changes in go modules, so we can see result of CI.
Did you mean temporary updating mod.go with pmm replaced with my fork or with version after https://github.com/percona/pmm/pull/863 being merged?

**Comment by BupycHuk** [comment]
@pkadej yes, updated go.mod with `replace` directive for now and once PR in `pmm` will be merged update it to that.

**Comment by pkadej** [comment]
> @pkadej yes, updated go.mod with `replace` directive for now and once PR in `pmm` is merged update go.mod to use that version.
@BupycHuk 
Done. Unfortunately i don't have permission to trigger runflow.

File: services/agents/channel/channel_test.go [hunk 6]
```diff
@@ -356,3 +356,50 @@ func TestUnexpectedResponsePayloadFromAgent(t *testing.T) {
 	close(stopServer)
 	<-stop
 }
+
+func TestChannelForDefaultsFileParser(t *testing.T) {
```
Comment (BupycHuk) [inline services/agents/channel/channel_test.go:360]: I'm not sure what are we testing here

**Review by BupycHuk — APPROVED** [review]

File: services/agents/parse_defaults_file.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+func createRequest(configPath string, serviceType models.ServiceType) (*agentpb.ParseDefaultsFileRequest, error) {
+	if serviceType == models.MySQLServiceType {
+		request := &agentpb.ParseDefaultsFileRequest{
+			ServiceType: inventorypb.ServiceType_MYSQL_SERVICE,
+			ConfigPath:  configPath,
+		}
+		return request, nil
+	} else {
+		return nil, errors.Errorf("unhandled service type %s", serviceType)
```
Comment (ritbl) [inline services/agents/parse_defaults_file.go:96]:
```suggestion
		return nil, errors.Errorf("unsupported service type %s", serviceType)
```

File: services/agents/parse_defaults_file.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	start := time.Now()
+	defer func() {
+		if dur := time.Since(start); dur > 5*time.Second {
+			l.Warnf("ParseDefaultsFile took %s.", dur)
+		}
+	}()
```
Comment (ritbl) [inline services/agents/parse_defaults_file.go:57]: Is it realistic? Reading a file with few lines should not be a problem. And, I think, it should be handed in agent

File: services/agents/parse_defaults_file.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+	resp, err := pmmAgent.channel.SendAndWaitResponse(request)
+	if err != nil {
+		return nil, err
+	}
+
+	l.Infof("ParseDefaultsFile response from agent: %+v.", resp)
+	parserResponse, ok := resp.(*agentpb.ParseDefaultsFileResponse)
+	if !ok {
+		return nil, errors.New("wrong response from agent (not ParseDefaultsFileResponse model)")
```
Comment (ritbl) [inline services/agents/parse_defaults_file.go:73]: Can we add info regarding actual type?

File: services/management/mysql.go [hunk 10]
```diff
@@ -78,6 +83,35 @@ func (s *MySQLService) Add(ctx context.Context, req *managementpb.AddMySQLReques
...
 			return err
 		}
+
+		if req.DefaultsFile != "" {
+			result, err := s.dfp.ParseDefaultsFile(ctx, req.PmmAgentId, req.DefaultsFile, models.MySQLServiceType)
+			if err != nil {
+				return status.Error(codes.FailedPrecondition, fmt.Sprintf("Defaults file error: %s.", err))
+			}
+
+			// set username and password from parsed defaults file by agent
```
Comment (ritbl) [inline services/management/mysql.go:93]: nit: it would be good to know that value has been overridden, can we log.debug that?

**Review by ritbl — APPROVED** [review]

**Review by artemgavrilov — APPROVED** [review]
LGTM with comments

**Pull request merged**